### PR TITLE
implement - SDL 0219 explicit returned type from NSDictionary(Store) category

### PR DESCRIPTION
**Pull request by mvyrko**
Fixes https://github.com/smartdevicelink/sdl_ios/issues/1153

This PR is **[not ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
We already test all values

### Summary
Need to verify types that return from function, that fixed a big amount of crashes.

##### Enhancements
* Return only types that we expect

##### Bug Fixes
* Fixed bugs with unrecognized selector where instead of one type we get another
`-[NSNull count]: unrecognized selector sent to instance 0x1e99699d0`
`-[__NSDictionaryI firstObject]: unrecognized selector sent to instance 0x282447e80`
`-[__NSCFNumber isEqualToEnum:]: unrecognized selector sent to instance 0xecbbf7f37a5c3eb2`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by kshala-ford** [comment]
Hello @mvyrko. I like your approach in this PR. As it addresses the topic of type checking it might require a minor version increase so I recommend creating a separate issue. The first thing that comes into my mind are possible nullability issues as some properties are `nonnull`.

File: SmartDeviceLink/NSMutableDictionary+Store.h [hunk 1]
```diff
@@ -18,9 +18,10 @@ typedef NSString* SDLEnum SDL_SWIFT_ENUM;
 @interface NSDictionary (Store)
 
 - (void)sdl_setObject:(NSObject *)object forName:(SDLName)name;
-- (nullable id)sdl_objectForName:(SDLName)name;
+- (nullable SDLEnum)sdl_enumForName:(SDLName)name;
+- (nullable NSArray<SDLEnum> *)sdl_enumsForName:(SDLName)name;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/NSMutableDictionary+Store.h]: I don't think we need separate getters for enums as they all are strings. We should just return `[self sdl_objectsForName:name ofClass:NSString.class];`

File: SmartDeviceLink/NSMutableDictionary+Store.m [hunk 2]
```diff
@@ -24,30 +24,48 @@ - (nullable id)sdl_objectForName:(SDLName)name {
...
 
-- (NSArray *)sdl_objectsForName:(SDLName)name ofClass:(Class)classType {
-    NSArray *array = [self sdl_objectForName:name];
-    if ([array isEqual:[NSNull null]]) {
-        return [NSMutableArray array];
-    } else if (array.count < 1 || [array.firstObject isMemberOfClass:classType]) {
+- (nullable NSArray *)sdl_objectsForName:(SDLName)name ofClass:(Class)classType {
+    NSArray *array = [self sdl_objectForName:name ofClass:NSArray.class];
+    if (!array) {
+        return nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/NSMutableDictionary+Store.m]: What is the logic behind returning `nil` vs an empty array?

File: SmartDeviceLink/NSMutableDictionary+Store.h (hunk 1 above)
Comment (mvyrko) [reply SmartDeviceLink/NSMutableDictionary+Store.h]: In this case, you know that SDLEnum is NSString.

File: SmartDeviceLink/NSMutableDictionary+Store.m (hunk 2 above)
Comment (mvyrko) [reply SmartDeviceLink/NSMutableDictionary+Store.m]: empty array is broken current logic, you can see it if you change to [NSArray array] and run tests.

File: SmartDeviceLink/NSMutableDictionary+Store.m [hunk 3]
```diff
@@ -24,30 +24,48 @@ - (nullable id)sdl_objectForName:(SDLName)name {
...
 
+- (nullable SDLEnum)sdl_enumForName:(SDLName)name {
+    return [self sdl_objectForName:name ofClass:NSString.class];
+}
+
+- (nullable NSArray<SDLEnum> *)sdl_enumsForName:(SDLName)name {
+    return [self sdl_objectsForName:name ofClass:NSString.class];
+}
+
 - (nullable id)sdl_objectForName:(SDLName)name ofClass:(Class)classType {
```
Comment (NicoleYarroch) [inline SmartDeviceLink/NSMutableDictionary+Store.m]: Currently we don't have test cases to check getting data that has a different data type than the expected data type. Can you *NSMutableDictionary+StoreSpec.m* file to test the `sdl_objectForName: ofClass:` and `sdl_objectsForName:ofClass:` methods to make sure the right data type is being returned.

Comment (mvyrko) [reply SmartDeviceLink/NSMutableDictionary+Store.m]: @NicoleYarroch Implemented in 5ee2ac0

**Comment by mvyrko** [comment]
@jordynmackool Please, review PR.

**Comment by jordynmackool** [comment]
@mvyrko 
Please submit a PR against this md file: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0219-ios-check-type.md in the evolution repo. 

This PR can be used as an example: https://github.com/smartdevicelink/sdl_evolution/pull/593

**Comment by theresalech** [comment]
@mvyrko I've created an issue for the accepted Evolution Proposal for this feature: #1198.  Please ensure this PR is reflective of the final Evolution Proposal and let us know when it's ready for review.  Thanks!

**Comment by mvyrko** [comment]
@theresalech Thank you, will do

**Comment by mvyrko** [comment]
Close this one, new PR is https://github.com/smartdevicelink/sdl_ios/pull/1199

**Pull request closed without merging**